src/python/pants/engine/legacy/graph.py

```diff
@@ -360,24 +361,43 @@ def transitive_dependents_of_addresses(self, addresses):
         return result
 
 
-@dataclass(frozen=True)
+@frozen_after_init
+@dataclass
 class HydratedTarget:
     """A wrapper for a fully hydrated TargetAdaptor object.
 
+    Why have this type if all information is stored in the underlying TargetAdaptor? We need it for
+    the type-driven graph to work properly. A HydratedTarget is a generic wrapper around different
+    target classes, whereas a TargetAdaptor has different types like PythonTestsAdaptor and
+    PythonLibraryAdaptor. We need to be able to work with targets both generically and depending
+    on their target type (i.e. unions).
+
     Transitive graph walks collect ordered sets of TransitiveHydratedTargets which involve a huge
-    amount of hashing: we implement eq/hash via direct usage of an Address field to speed that up.
+    amount of hashing: we implement a custom eq/hash to speed that up.
     """
 
-    address: Address
     adaptor: TargetAdaptor
-    dependencies: Tuple[Address, ...]
 
-    @property
-    def addresses(self) -> Tuple:
-        return self.dependencies
+    def __init__(self, adaptor: TargetAdaptor) -> None:
+        self.adaptor = adaptor
+        # This field is set for efficient lookup of the address in `__hash__` and `__eq__` during
+        # graph walks. Directly accessing this field, rather than using `adaptor.address` is about
+        # 10x faster. However, rules should still use `adaptor.address` for clarity and because
+        # they do not need to access the address as frequently, so it would be an over-optimization
+        # (we're talking about 10^-7 vs. 10^-6 seconds here).
+        self._address = adaptor.address
```

Comment:
Directly using `adaptor.address` resulted in going from about 4.5 seconds to 10 seconds, although I only tested with 2 runs. 

I got these 10^-7 numbers by using `timeit(lambda: adaptor.address, number=10_000)` vs. `timeit(lambda: self._address, number=10_000)`.

Reply:
This implies that `__eq__` is being executed ~5.5*10^6 times in your test??

Reply:
Both `__eq__` and this rule:

https://github.com/pantsbuild/pants/blob/6a81a2d74ae38d60fd836bddb144acaf8419ac05/src/python/pants/engine/legacy/graph.py#L583-L596

Specifically, I applied this diff:

```
diff --git a/src/python/pants/engine/legacy/graph.py b/src/python/pants/engine/legacy/graph.py
index bfd5d31b3..d80a0afae 100644
--- a/src/python/pants/engine/legacy/graph.py
+++ b/src/python/pants/engine/legacy/graph.py
@@ -255,9 +255,13 @@ class LegacyBuildGraph(BuildGraph):
 
         logger.debug("Injecting address specs to %s: %s", self, address_specs)
         with self._resolve_context():
+            from time import time
+            start = time()
             (thts,) = self._scheduler.product_request(
                 LegacyTransitiveHydratedTargets, [address_specs]
             )
+            print(round(time() - start, 2))
+            raise SystemExit
 
         self._index(thts.closure)
 
@@ -385,17 +389,17 @@ class HydratedTarget:
         # 10x faster. However, rules should still use `adaptor.address` for clarity and because
         # they do not need to access the address as frequently, so it would be an over-optimization
         # (we're talking about 10^-7 vs. 10^-6 seconds here).
-        self._address = adaptor.address
+        # self._address = adaptor.address
 
     def __hash__(self):
-        return hash((self._address, self.adaptor))
+        return hash(self.adaptor)
 
     def __eq__(self, other):
         if not isinstance(other, HydratedTarget):
             return NotImplemented
         # NB: This short-circuiting is essential. Without it, constructing the build graph for the
         # Pants repository takes 71 seconds compared to 4.3 seconds!
-        if self._address != other._address:
+        if self.adaptor.address != other.adaptor.address:
             return False
         return self.adaptor == other.adaptor
 
@@ -586,9 +590,9 @@ async def legacy_transitive_hydrated_targets(
     addresses: Addresses,
 ) -> LegacyTransitiveHydratedTargets:
     thts = await Get[TransitiveHydratedTargets](Addresses, addresses)
-    roots_bfas = await MultiGet(Get[BuildFileAddress](Address, ht._address) for ht in thts.roots)
+    roots_bfas = await MultiGet(Get[BuildFileAddress](Address, ht.adaptor.address) for ht in thts.roots)
     closure_bfas = await MultiGet(
-        Get[BuildFileAddress](Address, ht._address) for ht in thts.closure
+        Get[BuildFileAddress](Address, ht.adaptor.address) for ht in thts.closure
     )
     return LegacyTransitiveHydratedTargets(
         roots=tuple(
```

Then, running the script from https://github.com/pantsbuild/pants/pull/9219#pullrequestreview-368500641 I got an average time of 11.0 seconds over 40 runs... 🤷‍♂️

My instinct was "wait, that can't be right!" The only alternative hypothesis I could come up with is that for some reason `self._address != self.adaptor.address`, which doesn't make sense the way the constructor is set up nor is it happening in reality - I added an assert that they're equivalent for every usage and indeed that's true.